### PR TITLE
New: Add google sheets support to new UI (and validation fix)

### PR DESCRIPTION
**Pull request by ErezBiren**

### CODE REVIEW

**Comment by baruchiro** [comment]
@ErezBiren I think your PR here includes the code for #427

File: ui-react/src/types.tsx [hunk 1]
```diff
@@ -43,10 +46,41 @@ export interface JsonConfig extends OutputVendorConfigBase {
   }
 }
 
+export interface Credentials {
```
Comment (brafdlog) [inline ui-react/src/types.tsx:49]:
```suggestion
export interface GoogleSheetsCredentials {
```

Comment (brafdlog) [inline ui-react/src/types.tsx:49]: Is this copied from somewhere? If so, can it be imported instead?

File: ui-react/src/eventsBridge.tsx [hunk 2]
```diff
@@ -51,3 +51,19 @@ export async function sentryUserReportProblem(reportProblem) {
 export async function sourceCommitShort() {
   return electron.ipcRenderer.invoke('sourceCommitShort');
 }
+
+export async function googleLogin() {
+  return electron.ipcRenderer.invoke('electronGoogleLogin');
+}
+
+export async function validateToken(credentials) {
```
Comment (brafdlog) [inline ui-react/src/eventsBridge.tsx:59]: validateGoogleToken

File: src/handlers/googleAuthHandlers.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import ElectronGoogleOAuth2 from 'electron-google-oauth2';
+import {
+  clientId, clientSecret, createClient, redirectUri, scopes, validateToken
+} from '@/backend/export/outputVendors/googleSheets/googleAuth';
+import { getAllSpreadsheets } from '@/backend/export/outputVendors/googleSheets/googleSheetsInternalAPI';
+import { createSpreadsheet } from '@/backend/export/outputVendors/googleSheets/googleSheets';
+
+export const googleLoginHandler = () => {
```
Comment (brafdlog) [inline src/handlers/googleAuthHandlers.ts:8]: I see there are almost no type definitions in the code.
It is very helpful (especially in a collaborative project) to have type definitions where possible. Please add for function parameters and return values of functions

File: src/handlers/index.ts [hunk 4]
```diff
@@ -19,7 +22,12 @@ const functions = {
   getConfig: getConfigHandler,
   updateConfig: updateConfigHandler,
   getYnabAccountData,
+  electronGoogleLogin: googleLoginHandler,
```
Comment (brafdlog) [inline src/handlers/index.ts:25]: Having function names consistently as verbs makes the code more readable.
So I suggest. electronGoogleLogin -> loginToGoogle / initiateGoogleLogin etc

File: src/handlers/index.ts [hunk 5]
```diff
@@ -19,7 +22,12 @@ const functions = {
...
+  validateToken: validateTokenHandler,
```
Comment (brafdlog) [inline src/handlers/index.ts:26]: Token is a generic term and can be relevant to other things. validateToken  -> validateGoogleToken

File: ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  const [readyToSave, setReadyToSave] = useState(false);
+
+  // const dataToReturn = setupExporterConfigForm(OutputVendorName.GOOGLE_SHEETS);
+  const [exporterConfig, setExporterConfig] = useState<GoogleSheetsConfig>(
+    exporter as GoogleSheetsConfig
+  );
+
+  useEffect(() => {
+    const validate = async () => {
+      // const valid = await validateToken(exporterConfig?.options?.credentials);
```
Comment (brafdlog) [inline ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx:47]: The logic here is commented out, so not sure what the whole use effect does

File: ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+      const credentials = await googleLogin();
+      setExporterConfig((prevExport) => ({
+        ...prevExport,
+        options: { ...prevExport.options, credentials },
+      }));
+
+      setStatus(Status.LOGGED_IN);
+    } catch (ex) {
+      setStatus(Status.ERROR);
+      SetErrorMessage(ex.message ?? ex);
```
Comment (brafdlog) [inline ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx:72]: We should log the exception itself to get the stack trace. Here you will get only the message and lose the stack

File: ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    }
+  };
+
+  const save = async () => {
+    const isNewSheet = exporterConfig.options.spreadsheetId.length < 30;
```
Comment (brafdlog) [inline ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx:77]: How does this mean that it is a new sheet?
I suggest adding a comment why length < 30 means new sheet

File: ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        exporterConfig.options.credentials
+      );
+      setExporterConfig((prevExport) => ({
+        ...prevExport,
+        options: { ...prevExport.options, spreadsheetId },
+      }));
+
+      setStatus(Status.LOGGED_IN);
+    }
+    // return dataToReturn.submit();
```
Comment (brafdlog) [inline ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx:91]: commented out code

File: ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+                >
+                  Save
+                </Button>
+              </Form>
+            </Card.Body>
+          </Card>
+        </>
+      )}
+    </div>
+    // <p>ההגדרות של גוגל שיטס עוד לא זמינות בממשק החדש. נא לעבור לממשק הישן כדי להגדיר</p>
```
Comment (brafdlog) [inline ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx:140]: commented out code

File: ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+function EditGoogleSheetsExporter({
+  exporter,
+  handleSave,
+}: EditGoogleSheetsExporterProps) {
+  const [status, setStatus] = useState<Status>(Status.LOADING);
+  const [errorMessage, SetErrorMessage] = useState();
+  const [readyToSave, setReadyToSave] = useState(false);
+
+  // const dataToReturn = setupExporterConfigForm(OutputVendorName.GOOGLE_SHEETS);
```
Comment (brafdlog) [inline ui-react/src/components/exporters/EditGoogleSheetsExporter.tsx:40]: commented out code

**Review by brafdlog — CHANGES_REQUESTED** [review]
Thanks Erez!
See comments

**Comment by ErezBiren** [comment]
@baruchiro -  Oops, you are right ... I accidentally added the google sheets  export code

**Comment by ErezBiren** [comment]
@brafdlog - Sorry for the premature code,
The google sheet code was added unintentionally.

**Pull request closed without merging**